src/db/collections/protocol_update.rs

```diff
@@ -73,36 +77,26 @@ impl ProtocolUpdateCollection {
         self.find_one(doc! { "parameters.version": version as i32 }, None).await
     }
 
-    /// Inserts a protocol parameters for a given milestone index.
-    pub async fn insert_protocol_parameters(
-        &self,
-        tangle_index: MilestoneIndex,
-        parameters: ProtocolParameters,
-    ) -> Result<(), Error> {
-        self.insert_one(
-            ProtocolUpdateDocument {
-                tangle_index,
-                parameters,
-            },
-            None,
-        )
-        .await?;
-
-        Ok(())
-    }
-
     /// Add the protocol parameters to the list if the protocol parameters have changed.
-    pub async fn update_latest_protocol_parameters(
+    pub async fn upsert_protocol_parameters(
         &self,
-        tangle_index: MilestoneIndex,
+        ledger_index: MilestoneIndex,
         parameters: ProtocolParameters,
     ) -> Result<(), Error> {
-        if let Some(latest_params) = self.get_latest_protocol_parameters().await? {
-            if latest_params.parameters != parameters {
-                self.insert_protocol_parameters(tangle_index, parameters).await?;
-            }
-        } else {
-            self.insert_protocol_parameters(tangle_index, parameters).await?;
+        let params = self.get_protocol_parameters_for_ledger_index(ledger_index).await?;
+        if params.is_none()
+            || params
+                .map(|latest_params| latest_params.parameters != parameters)
+                .unwrap_or_default()
+        {
+            self.update_one(
+                doc! { "_id": ledger_index },
+                doc! { "$set": {
+                    "parameters": mongodb::bson::to_bson(&parameters).unwrap()
+                } },
+                UpdateOptions::builder().upsert(true).build(),
+            )
+            .await?;
```

Comment:
here as well ... I think `!matches!` is still preferable ... and the undocumented `unwrap`